File: app/src/main/java/me/ykrank/s1next/data/api/model/HomeReply.java
package me.ykrank.s1next.data.api.model;

import android.support.annotation.Nullable;

import com.github.ykrank.androidtools.guava.Objects;
import com.github.ykrank.androidtools.ui.adapter.model.SameItem;
import com.github.ykrank.androidtools.util.L;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import me.ykrank.s1next.data.api.model.wrapper.HomeReplyWebWrapper;

/**
 * Created by ykrank on 2017/2/5.
 * User's reply model
 */

public class HomeReply implements SameItem, HomeReplyWebWrapper.HomeReplyItem {
    private String reply;
    //eg forum.php?mod=redirect&goto=findpost&ptid=1220112&pid=34645514
    private String url;

    @Nullable
    public static HomeReply fromHtmlElement(Element element) {
        HomeReply reply = null;
        try {
            Elements tdReplies = element.children();
            if (tdReplies.size() < 1) {
                return null;
            }
            Element tdReply = tdReplies.first();
            Elements elesReply = tdReply.children();
            if (elesReply.size() < 3) {
                return null;
            }
            reply = new HomeReply();
            //reply
            Element eleReply = elesReply.get(1);
            reply.setReply(eleReply.text());
            //eg thread-1220112-1-1.html
            reply.setUrl(eleReply.attr("href"));
        } catch (Exception e) {
            L.report(new RuntimeException(element.toString(), e));
        }
        return reply;
    }

    public String getReply() {
        return reply;
    }

    public void setReply(String reply) {
        this.reply = reply;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        HomeReply homeReply = (HomeReply) o;
        return Objects.equal(reply, homeReply.reply) &&
                Objects.equal(url, homeReply.url);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(reply, url);
    }

    @Override
    public String toString() {
        return "HomeReply{" +
                "reply='" + reply + '\'' +
                ", url='" + url + '\'' +
                '}';
    }

    @Override
    public boolean isSameItem(Object o) {
        if (this == o) return true;
        if (!(o instanceof HomeReply)) return false;
        HomeReply that = (HomeReply) o;
        return Objects.equal(url, that.url);
    }
}

File: app/src/main/java/me/ykrank/s1next/data/api/model/HomeReply.kt
package me.ykrank.s1next.data.api.model

import com.github.ykrank.androidtools.ui.adapter.StableIdModel
import com.github.ykrank.androidtools.ui.adapter.model.DiffSameItem
import com.github.ykrank.androidtools.util.L
import me.ykrank.s1next.data.api.model.wrapper.HomeReplyWebWrapper
import org.jsoup.nodes.Element

/**
 * Created by ykrank on 2017/2/5.
 * User's reply model
 */

class HomeReply : DiffSameItem, HomeReplyWebWrapper.HomeReplyItem, StableIdModel {
    var reply: String? = null
    //eg forum.php?mod=redirect&goto=findpost&ptid=1220112&pid=34645514
    var url: String? = null



    override fun toString(): String {
        return "HomeReply{" +
                "reply='" + reply + '\''.toString() +
                ", url='" + url + '\''.toString() +
                '}'.toString()
    }

    override fun isSameItem(other: Any): Boolean {
        if (this === other) return true
        if (javaClass != other.javaClass) return false

        other as HomeReply

        if (url != other.url) return false

        return true
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as HomeReply

        if (reply != other.reply) return false
        if (url != other.url) return false

        return true
    }

    override fun hashCode(): Int {
        var result = reply?.hashCode() ?: 0
        result = 31 * result + (url?.hashCode() ?: 0)
        return result
    }

    companion object {

        fun fromHtmlElement(element: Element): HomeReply? {
            var reply: HomeReply? = null
            try {
                val tdReply = element.children().first() ?: return null
                val elesReply = tdReply.children()
                if (elesReply.size < 3) {
                    return null
                }
                reply = HomeReply()
                //reply
                val eleReply = elesReply[1]
                reply.reply = eleReply.text()
                //eg thread-1220112-1-1.html
                reply.url = eleReply.attr("href")
            } catch (e: Exception) {
                L.report(RuntimeException(element.toString(), e))
            }

            return reply
        }
    }
}
